Backend security #233

romanstetsyk opened the pull request:
# Description

Add error handler
Add 404 handler
Update event controllers (getOne, getAll, deleteEvent, deleteAllEvents, create).
Add middleware to validate the request body before inserting data into the database.
For now, I only implemented some date validation (the start date can't be in the past, the difference between the end date and start date can be 3 months maximum).

## Type of change

Please select everything applicable. Please, do not delete any lines.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Issue
- [x] Is this related to a specific issue? Is so please specify: #233 #232 

# Checklist:

- [x] I have executed `npm run lint` and resolved any outstanding errors. Most issues can be solved by executing `npm run format`
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

romanstetsyk commented:
Requesting a code review

Caleb-Cohen commented on server/server.js:
```diff
@@ -38,7 +41,7 @@ app.use(
 
 // Render React as View
-app.use(express.static(path.join(__dirname, "..", "client", "build")));
+// app.use(express.static(path.join(__dirname, "..", "client", "build")));
```
Any concerns with this breaking production build?

romanstetsyk replied:
Honestly, I'm not sure. It works on my end. The idea is that the server serves JSON responses only, not views. I might be wrong though

timmyichen replied:
In production, the server serves both.  The client is only a separate app in development; in production, you compile the client app into some static assets, then serve the static assets with the middleware above. This probably breaks production :P

romanstetsyk replied:
Oh no, I am going to revert this too!

Is this how monorepos work? I only dealt with separate server and client repos, where the server responds with JSON and the client consumes it.

timmyichen replied:
Separate server and client repos are a common pattern in development, but you'll probably never see this used in production.

Caleb-Cohen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
     }
   },
-  getAll: async (req, res) => {
+  getAll: async (req, res, next) => {
```
What is the next parameter accomplishing?

romanstetsyk replied:
Passes the request to the next controller. i use it here mostly in the `catch` block to pass the error to the error handler

Caleb-Cohen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+
+      res.sendStatus(204);
+    } catch (err) {
+      next(err);
```
Excellent clean up here. Great work.

Caleb-Cohen commented on server/server.js:
```diff
@@ -17,6 +17,9 @@ const eventsRoutes = require("./routes/events");
 require("./config/passport")(passport);
 
+// Mongoose debug mode
+// mongoose.set("debug", true);
```
Can this go?

Caleb-Cohen reviewed (commented):
Left some basic comments.

timmyichen commented on server/config/passport.js:
```diff
@@ -1,13 +1,19 @@
-const DiscordStrategy = require("passport-discord").Strategy;
+const DiscordStrategy = require("passport-discord");
```
This seems like a breaking change?  Is the default export of `passport-discord` also a named export `Strategy`?

romanstetsyk replied:
I messed up, but it's still working for some reason. Probably some extension interferes with imports. Will fix it

timmyichen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+  getOne: async (req, res, next) => {
     try {
-      const event = await Event.findById(req.params.id);
+      const { id } = req.params;
+
+      // Check if the ID is valid
+      if (!mongoose.isValidObjectId(id)) {
+        throw httpError(404);
+      }
```
This can be abstracted into a middleware that you can use on each of these routes - not a blocking comment, just a nice-to-have. :)

romanstetsyk replied:
Good point! Will do

timmyichen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+      }).exec();
+
+      if (count === 0) {
+        throw httpError(401);
```
I think that a 404 here would be more appropriate.

> The HyperText Transfer Protocol (HTTP) 401 Unauthorized response status code indicates that the client request has not been completed because it lacks valid authentication credentials for the requested resource.

It's not that they lack valid authentication; this route's purpose is "delete all events in this group", which, if the group doesn't exist, then they're trying to delete something that doesn't exist.  Not found!

romanstetsyk replied:
I've updated it to 404.
This is from the document I linked in another comment:
<img width="1059" alt="image" src="https://user-images.githubusercontent.com/25715951/212154259-26ee9a34-2d05-4927-a294-422750890f60.png">

Do you think it's a good idea to remove the query that counts the documents and only leave the `Event.deleteMany` query?

timmyichen replied:
Yeah - we can basically trust `deleteMany` to delete all of the 

And re the status code - I think that source misunderstand what, exactly, is meant to be idempotent.  This stackoverflow answer explains it pretty well, I think: https://stackoverflow.com/questions/4088350/is-rest-delete-really-idempotent

timmyichen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+      // If the number of documents found is not equal to the number of deleted documents
+      // Something may have gone wrong
+      if (count !== deletedCount) {
+        console.log(`Documents found: ${count}`);
+        console.log(`Documents deleted: ${deletedCount}`);
```
Can these be consolidated into the same line?

timmyichen commented on server/config/passport.js:
```diff
@@ -1,13 +1,19 @@
 
-  passport.deserializeUser((id, done) => {
-    User.findById(id, (err, user) => done(err, user));
+  passport.deserializeUser(async (id, done) => {
+    try {
+      const user = await User.findById(id);
+      if (!user) throw new Error("User not found");
+      done(null, user);
+    } catch (err) {
+      done(err);
+    }
```
what is this doing that the previous code isn't doing?

romanstetsyk replied:
Sometimes I get errors from this function that stops the server. I thought that it was because `User.findById` is asynchronous, but `(id, done) => {...` doesn't have `async`?

timmyichen replied:
Both functions basically are dependent on `done` being called, so regardless of whether it's async or sync, it "resolves" when `done` is called.  I think you could actually do something as simple as:

```js
passport.deserializeUser((id, done) => {
    User.findById(id, done);
})
```

Or just (for the same reason):

```js
passport.deserializeUser(User.findById)
```

This is much less readable though so the original version is probably preferred over these code-golfy ones 🙃

romanstetsyk replied:
I'll change the code back to what it was and will let you know when it occurs again. I think it was in postman, not in the browser.

timmyichen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+  getAll: async (req, res, next) => {
     try {
-      const events = await Event.find().populate('user').exec();
+      // Get an array of events
+      const events = await Event.find()
+        .populate("user", "displayName")
+        .lean()
+        .exec();
+
       // return all events
       res.json(events);
     } catch (err) {
-      console.log(err);
+      next(err);
     }
```
I'm not a big fan of the "wrap everything in a try block and catch errors to pass to the error handler".  If the goal is to get asynchronous controller functions' errors to be caught by the the error handler, it's probably cleaner to use a package like [express-async-router](https://www.npmjs.com/package/express-async-router#promise-handling) instead.

The general idea is that try/catch should be as surgical as possible - we should be wrapping code with try/catch blocks when we _expect_ that they could error, and let the error handler take care of any unexpected errors.  Otherwise it's difficult to distinguish between errors that we expect to happen, and errors that catch us by surprise.

romanstetsyk replied:
Thanks for the feedback! I'll look into that package. Would you happen to have an example of a repo that uses it?

> let the error handler take care of any unexpected errors

If you don't wrap everything into try/catch, how is the error passed to the error handler? Wouldn't it just stop the server?

timmyichen replied:
> Thanks for the feedback! I'll look into that package. Would you happen to have an example of a repo that uses it?

I might have an example in one of mine, but I think the [usage](https://www.npmjs.com/package/express-async-router#usage) is pretty straightforward.  [This line](https://github.com/Caleb-Cohen/Together/pull/244/files/04956ef7d6bc0de702e05529a943a88969ba1118#diff-2ccfbef22d80ff5b83272c1bbdeab560e418d3b600526c7f57267014433d4e03R2) would essentially be replaced with the `router` declaration from the docs.

> If you don't wrap everything into try/catch, how is the error passed to the error handler? Wouldn't it just stop the server?

Not at all! See [Express Error Handling](https://expressjs.com/en/guide/error-handling.html), specifically:

> Errors that occur in synchronous code inside route handlers and middleware require no extra work. If synchronous code throws an error, then Express will catch and process it.

This is for synchronous code - however, later down the page:

> You must catch errors that occur in asynchronous code invoked by route handlers or middleware and pass them to Express for processing.

This is less idea for the reasons I described above.  Luckily, folks wrote a package to handle async errors in the same way, so that we can keep the same patterns for our controllers no matter whether they are sync or async.

timmyichen reviewed (commented):
Nice work!  Any particular reason you did not want to go with mongoose's [built in validators](https://mongoosejs.com/docs/validation.html)? 

Additionally, the linked issue seems to specify some additional pieces not covered by this PR; do you plan on adding it to this PR or following up in additional PRs?

timmyichen commented on server/controllers/events.js:
```diff
@@ -1,61 +1,107 @@
+      // Prevent users that are authenticated from deleting events they do not author.
+      const event = await Event.findOne({ _id: id, user: req.user._id });
+      if (!event) {
+        throw httpError(401);
```
I think [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) has pretty succinct descriptions of each!

timmyichen replied:
I guess in this particular case, you want to do two things:
* Checking to see if the event ID exists, and (404 if doesnt exist)
* Check to see if the event is owned by the user (403 if they dont)

You _could_ run two different queries to determine which status code to send back.  In an app where every record (event) isn't publicly accessible, differentiating between the two situations is actually _detrimental_, since if you get a 403, that means that the resource with that ID exists, and then you can try probing for other vulnerabilities by passing that ID in other routes.  For this reason, we generally want to use 404 as a "this doesn't exist _to you_"

romanstetsyk commented:
@timmyichen It was a friend of mine that recommended using `Joi`. `Joi` and `mongoose` validation technically serve different purposes and complement each other. Just another layer of security. `Joi` is used to validate a request body, for example, sent in a `POST` request. So you can receive a form data, validate it with `Joi`, transform it, and submit it to the database. Mongoose will validate it before sending it to DB.

romanstetsyk commented:
Hey @timmyichen, could you do a code review? I ended up using the `express-async-errors` package to handle async errors without using `try/catch` (simply because it has more downloads on npm, 236k vs 12k), added a middleware to check if the number of events is 5, and updated the mongoose schema.

timmyichen commented:
> Hey @timmyichen, could you do a code review? I ended up using the `express-async-errors` package to handle async errors without using `try/catch` (simply because it has more downloads on npm, 236k vs 12k), added a middleware to check if the number of events is 5, and updated the mongoose schema.

Would it be possible to resolve merge conflicts before I review? Also, if this is ready for review feel free to kick it out of draft mode :)

romanstetsyk commented:
> Would it be possible to resolve merge conflicts before I review? Also, if this is ready for review feel free to kick it out of draft mode :)

I didn't realize the development branch was updated! It's ready

timmyichen reviewed (approved):
Code changes LGTM, though i haven't tested it locally

Caleb-Cohen commented:
<img width="495" alt="Screenshot 2023-01-18 at 6 39 39 PM" src="https://user-images.githubusercontent.com/75268174/213328768-ae7120d3-afc1-4e1d-ae4a-a7c77c14fd2c.png">
<img width="944" alt="Screenshot 2023-01-18 at 6 40 02 PM" src="https://user-images.githubusercontent.com/75268174/213328771-9cfa2ab3-756b-4728-bbc4-4c51f511a337.png">

Any idea why I'm hitting this error?

Caleb-Cohen commented:
> It looks like there's a merge conflict with the latest `development` update

There are, but these errors are just from checking out your PR pre-merge conflicts.

romanstetsyk commented:
Have you installed the `express-async-errors` package? I've added it in one of the last commits

Caleb-Cohen commented:
> Have you installed the `express-async-errors` package? I've added it in one of the last commits

I've done `npm install` twice - anything else I should be doing?
<img width="402" alt="Screenshot 2023-01-18 at 7 03 23 PM" src="https://user-images.githubusercontent.com/75268174/213331312-6593215e-d341-4f98-b0b8-a7d9ca9336bb.png">
